### PR TITLE
Hook up IR_FLAG_NOTHROW

### PR DESCRIPTION
**Pull request by Keno**
I added the flag in #45272, but didn't hook it up to the nothrow
model. In the fullness of time, these should forwarded from
inference and only recomputed if necessary, but for the moment,
just make it work.

### CODE REVIEW

File: base/compiler/ssair/ir.jl [hunk 1]
```diff
@@ -250,7 +250,7 @@ function setindex!(is::InstructionStream, newval::Instruction, idx::Int)
     is.flag[idx] = newval[:flag]
     return is
 end
-function setindex!(is::InstructionStream, newval::AnySSAValue, idx::Int)
+function setindex!(is::InstructionStream, newval::Union{AnySSAValue, Nothing}, idx::Int)
```
Comment (aviatesk) [inline base/compiler/ssair/ir.jl:253]: This change seems unnecessary?

File: base/compiler/ssair/ir.jl [hunk 2]
```diff
@@ -343,7 +343,7 @@ function getindex(x::IRCode, s::SSAValue)
     end
 end
 
-function setindex!(x::IRCode, repl::Union{Instruction, AnySSAValue}, s::SSAValue)
+function setindex!(x::IRCode, repl::Union{Instruction, Nothing, AnySSAValue}, s::SSAValue)
```
Comment (aviatesk) [inline base/compiler/ssair/ir.jl:346]: Same as above.

File: base/compiler/optimize.jl [hunk 3]
```diff
@@ -188,81 +188,92 @@ function stmt_affects_purity(@nospecialize(stmt), ir)
...
+    # TODO: We're duplicating analysis from inference here.
+    isa(stmt, PiNode) && return (true, true)
+    isa(stmt, PhiNode) && return (true, true)
+    isa(stmt, ReturnNode) && return (false, true)
+    isa(stmt, GotoNode) && return (false, true)
+    isa(stmt, GotoIfNot) && return (false, argextype(stmt.cond, src) ⊑ Bool)
+    isa(stmt, Slot) && return (false, false) # Slots shouldn't occur in the IR at this point, but let's be defensive here
+    if isa(stmt, GlobalRef)
+        nothrow = isdefined(stmt.mod, stmt.name)
+        return (nothrow, nothrow)
```
Comment (aviatesk) [inline base/compiler/optimize.jl:207]:
```suggestion
        return (nothrow && isconst(stmt.mod, stmt.name), nothrow)
```

File: base/compiler/optimize.jl [hunk 4]
```diff
@@ -188,81 +188,92 @@ function stmt_affects_purity(@nospecialize(stmt), ir)
...
-            fieldcount(typ) >= length(args) - 1 || return false
+            fieldcount(typ) >= length(args) - 1 || return (false, false)
             for fld_idx in 1:(length(args) - 1)
                 eT = argextype(args[fld_idx + 1], src)
                 fT = fieldtype(typ, fld_idx)
-                eT ⊑ fT || return false
+                eT ⊑ fT || return (false, false)
             end
-            return true
+            return (true, true)
```
Comment (aviatesk) [inline base/compiler/optimize.jl:252]:
```suggestion
            return (!ismutabletype(typ), true)
```

File: base/compiler/optimize.jl [hunk 5]
```diff
@@ -188,81 +188,92 @@ function stmt_affects_purity(@nospecialize(stmt), ir)
 end
 
 """
-    stmt_effect_free(stmt, rt, src::Union{IRCode,IncrementalCompact})
+    stmt_effect_flags(stmt, rt, src::Union{IRCode,IncrementalCompact})
+
+Returns a tuple of (total, nothrow) for a given statement.
 
 Determine whether a `stmt` is "side-effect-free", i.e. may be removed if it has no uses.
```
Comment (aviatesk) [inline base/compiler/optimize.jl:195]:
```suggestion
Returns a tuple of (total, nothrow) for a given statement.
```

File: base/compiler/ssair/ir.jl (hunk 1 above)
Comment (Keno) [reply base/compiler/ssair/ir.jl:253]: This regressed at some point. It's the idiom for deleting and instruction so should work, but yes, it's mostly unrelated to this PR.

File: base/compiler/optimize.jl (hunk 3 above)
Comment (Keno) [reply base/compiler/optimize.jl:207]: This should keep doing what it did before, but probably the `total` in the comment is from and it should actually be effect_free. This code predates the clarification of those terms.

**Comment by vtjnash** [comment]
sgtm

**Pull request merged**